runtime/common/QuditIdTracker.h

```diff
@@ -0,0 +1,65 @@
+/*************************************************************** -*- C++ -*- ***
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ *******************************************************************************/
+
+#pragma once
+
+#include <algorithm>
+#include <deque>
+
+namespace cudaq {
+
+/// @brief The QuditIdTracker tracks unique integer
+/// indices for allocated qudits. It will recycle indices
+/// when the qudit is deallocated, so that it can be reused.
+class QuditIdTracker {
+private:
+  /// @brief The current identifier.
+  std::size_t currentId = 0;
+
+  /// @brief The queue of recycled qubits.
+  std::deque<std::size_t> recycledQudits;
```

Comment:
Do we really need a `std::deque` here? Since the `returnIndex` methods sort the queue, we might as well just use a `std::vector` and pop the last one.

Reply:
That is a much better idea... 😄